File: Swift/Leetcode.playground/Sources/11_BinaryTreeGeneral/LC_549_BinaryTreeLongestConsecutiveSequenceII.swift
//
//  LC_549_BinaryTreeLongestConsecutiveSequenceII.swift
//  Leetcode
//
//  Created by viktyz on 2023/8/13.
//

import Foundation

public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init() { self.val = 0; self.left = nil; self.right = nil; }
    public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
    public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
        self.val = val
        self.left = left
        self.right = right
    }
}

public class Solution_549 {
    
    public var maxValue : Int = 0;
    
    public func longestConsecutive(_ root: TreeNode?) -> Int {
        longestPath(root)
        return maxValue
    }

    public func longestPath(_ root: TreeNode?) -> [Int] {
        if (root == nil) {
            return [0,0]
        }
        
        var increase : Int = 1;
        var decrease : Int = 1;
        
        if let left = root?.left {
            let L : [Int] = longestPath(left)
            
            if(root?.val == left.val + 1){
                decrease = L[1] + 1;
            }
            else if(root?.val == left.val - 1){
                increase = L[0] + 1;
            }
        }
        
        if let right = root?.right {
            let R : [Int] = longestPath(right)
            
            if(root?.val == right.val + 1){
                decrease = max(decrease,R[1] + 1);
            }
            else if(root?.val == right.val - 1){
                increase = max(increase,R[0] + 1);
            }
        }
        
        maxValue = max(maxValue,decrease + increase - 1)
        
        return [increase,decrease]
    }
}

public class LC_549_BinaryTreeLongestConsecutiveSequenceII : Solution{
    public override init() {}
}

File: JavaScript/11_BinaryTreeGeneral/LC_549_BinaryTreeLongestConsecutiveSequenceII.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestConsecutive = function(root) {

    var maxValue = 0;

    function longestPath(root){
        if(!root){
            return [0,0];
        }

        var increase = 1;
        var decrease = 1;

        if(root.left){
            let L = longestPath(root.left);
            if(root.val == root.left.val + 1){
                decrease = L[1] + 1; //左子节点值比当前值大 1，与当前值构成降序
            }
            else if(root.val == root.left.val - 1){
                increase = L[0] + 1; //左子节点值比当前值小 1，与当前值构成升序
            }
        }

        if(root.right){
            let R = longestPath(root.right);
            if(root.val == root.right.val + 1){
                decrease = Math.max(decrease, R[1] + 1); //右子节点值比当前值大 1，与当前值构成降序
            }
            else if(root.val == root.right.val - 1){
                increase = Math.max(increase, R[0] + 1); //右子节点值比当前值小 1，与当前值构成升序
            }
        }

        maxValue = Math.max(maxValue,decrease + increase - 1);

        return [increase,decrease];
    }

    longestPath(root);
    return maxValue;
};